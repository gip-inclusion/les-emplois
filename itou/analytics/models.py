import uuid

from django.db import models
from django.utils import timezone

from itou.common_apps.address.departments import DEPARTMENTS
from itou.users.enums import UserKind


class DatumCode(models.TextChoices):
    # Employee record - Base
    EMPLOYEE_RECORD_COUNT = "ER-001", "FS totales"
    EMPLOYEE_RECORD_DELETED = "ER-002", "FS (probablement) supprimées"
    # Employee record - Lifecycle
    EMPLOYEE_RECORD_PROCESSED_AT_FIRST_EXCHANGE = "ER-101", "FS intégrées (0000) au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_AT_FIRST_EXCHANGE = "ER-102", "FS avec une erreur au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_3436_AT_FIRST_EXCHANGE = "ER-102-3436", "FS avec une erreur 3436 au premier retour"
    EMPLOYEE_RECORD_WITH_AT_LEAST_ONE_ERROR = "ER-103", "FS ayant eu au moins un retour en erreur"
    # Approval - Base
    APPROVAL_COUNT = "AP-001", "PASS IAE total"
    APPROVAL_CANCELLED = "AP-002", "PASS IAE annulés"
    # Approval - PE notification cycle
    APPROVAL_PE_NOTIFY_SUCCESS = "AP-101", "PASS IAE synchronisés avec succès avec pole emploi"
    APPROVAL_PE_NOTIFY_PENDING = "AP-102", "PASS IAE en attente de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_ERROR = "AP-103", "PASS IAE en erreur de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_READY = "AP-104", "PASS IAE prêts à être synchronisés avec pole emploi"
    # Users
    USER_COUNT = "US-001", "Nombre d'utilisateurs"
    USER_JOB_SEEKER_COUNT = "US-011", "Nombre de demandeurs d'emploi"
    USER_PRESCRIBER_COUNT = "US-012", "Nombre de prescripteurs"
    USER_EMPLOYER_COUNT = "US-013", "Nombre d'employeurs"
    USER_LABOR_INSPECTOR_COUNT = "US-014", "Nombre d'inspecteurs du travail"
    USER_ITOU_STAFF_COUNT = "US-015", "Nombre d'administrateurs"
    # API usage
    API_TOTAL_CALLS = "API-001", "API : total d'appels reçus"
    API_TOTAL_UV = "API-002", "API : total de visiteurs uniques"
    API_TOTAL_CALLS_CANDIDATS = "API-003", "API candidats : total d'appels reçus"
    API_TOTAL_UV_CANDIDATS = "API-004", "API candidats : total de visiteurs uniques"
    API_TOTAL_CALLS_GEIQ = "API-005", "API GEIQ : total d'appels reçus"
    API_TOTAL_UV_GEIQ = "API-006", "API GEIQ : total de visiteurs uniques"
    API_TOTAL_CALLS_ER = "API-007", "API FS : total d'appels reçus"
    API_TOTAL_UV_ER = "API-008", "API FS : total de visiteurs uniques"
    API_TOTAL_CALLS_MARCHE = "API-009", "API Le marché : total d'appels reçus"
    API_TOTAL_CALLS_SIAES = "API-010", "API siaes : total d'appels reçus"
    API_TOTAL_UV_SIAES = "API-011", "API siaes : total de visiteurs uniques (par adresse IP)"
    API_TOTAL_CALLS_STRUCTURES = "API-012", "API structures : total d'appels reçus"


class Datum(models.Model):
    """Store an aggregated `value` of the `code` data point for the specified `bucket`."""

    id = models.UUIDField(primary_key=True, default=uuid.uuid4)

    code = models.TextField(choices=DatumCode.choices)
    bucket = models.TextField()
    value = models.IntegerField()  # Integer offers the best balance between range, storage size, and performance

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "data"
        unique_together = ["code", "bucket"]
        indexes = [models.Index(fields=["measured_at", "code"])]


class StatsDashboardVisit(models.Model):
    dashboard_id = models.IntegerField(verbose_name="ID tableau de bord Metabase")
    dashboard_name = models.TextField(verbose_name="nom de la vue du tableau de bord")
    department = models.CharField(verbose_name="département", choices=DEPARTMENTS.items(), max_length=3, null=True)
    region = models.TextField(verbose_name="région", null=True)
    current_company_id = models.IntegerField(verbose_name="ID entreprise courante", null=True)
    current_prescriber_organization_id = models.IntegerField(
        verbose_name="ID organisation prescriptrice courante", null=True
    )
    current_institution_id = models.IntegerField(verbose_name="ID institution courante", null=True)
    user_kind = models.TextField(verbose_name="type d'utilisateur", choices=UserKind.choices)
    user_id = models.IntegerField(verbose_name="ID utilisateur")

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "visite de tableau de bord"
