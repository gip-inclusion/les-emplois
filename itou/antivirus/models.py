from django.db import models
from django.utils.text import capfirst

from itou.files.models import File


class Scan(models.Model):
    file = models.OneToOneField(File, on_delete=models.CASCADE)
    clamav_signature = models.TextField()
    clamav_completed_at = models.DateTimeField(null=True, verbose_name="analyse ClamAV le")
    clamav_infected = models.BooleanField(null=True, verbose_name="fichier infect√© selon ClamAV")
    comment = models.TextField(blank=True, verbose_name="commentaire")

    class Meta:
        verbose_name = "analyse antivirus"
        verbose_name_plural = "analyses antivirus"

    def __str__(self):
        text = f"{capfirst(self._meta.verbose_name)} {self.file_id}"
        if self.clamav_infected:
            text = f"[VIRUS] {text}"
        return text
