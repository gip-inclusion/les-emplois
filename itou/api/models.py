import secrets

from django.db import models
from django.utils import timezone

from itou.companies.models import Company


def _generate_key():
    return secrets.token_urlsafe()


class CompanyToken(models.Model):
    key = models.CharField(default=_generate_key, unique=True, editable=False)
    label = models.CharField(verbose_name="m√©mo permettant d'identifier l'usage du jeton", max_length=60, unique=True)
    created_at = models.DateTimeField(default=timezone.now)
    companies = models.ManyToManyField(Company, related_name="api_tokens")

    class Meta:
        verbose_name = "jeton d'API entreprise"
        verbose_name_plural = "jetons d'API entreprise"
