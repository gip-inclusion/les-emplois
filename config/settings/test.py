import os


ITOU_ENVIRONMENT = "DEV"
os.environ["ITOU_ENVIRONMENT"] = ITOU_ENVIRONMENT

from .base import *  # noqa: E402,F403


SECRET_KEY = "foobar"
ALLOWED_HOSTS = []

# We *want* to do the same `collectstatic` on the CI than on PROD to catch errors early,
# but we don't want to do it when running the test suite locally for performance reasons.
if not os.getenv("CI", False):
    # `ManifestStaticFilesStorage` (used in base settings) requires `collectstatic` to be run.
    STORAGES["staticfiles"]["BACKEND"] = "django.contrib.staticfiles.storage.StaticFilesStorage"  # noqa: F405

ASP_ITOU_PREFIX = "XXXXX"  # same as in our fixtures
ITOU_PROTOCOL = "http"
ITOU_FQDN = "localhost:8000"

DATABASES["default"]["HOST"] = os.getenv("PGHOST", "127.0.0.1")  # noqa: F405
DATABASES["default"]["PORT"] = os.getenv("PGPORT", "5432")  # noqa: F405
DATABASES["default"]["NAME"] = os.getenv("PGDATABASE", "itou")  # noqa: F405
DATABASES["default"]["USER"] = os.getenv("PGUSER", "postgres")  # noqa: F405
DATABASES["default"]["PASSWORD"] = os.getenv("PGPASSWORD", "password")  # noqa: F405

MAILJET_API_KEY_PRINCIPAL = "API_MAILJET_KEY_PRINCIPAL"
MAILJET_SECRET_KEY_PRINCIPAL = "API_MAILJET_SECRET_PRINCIPAL"

AWS_S3_ENDPOINT_URL = f"http://{os.getenv('CELLAR_ADDON_HOST', 'localhost:9000')}/"
AWS_S3_ACCESS_KEY_ID = "minioadmin"
AWS_S3_SECRET_ACCESS_KEY = "minioadmin"
AWS_STORAGE_BUCKET_NAME = "tests"
