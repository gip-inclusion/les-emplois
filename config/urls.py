from django.conf import settings
from django.contrib import admin
from django.urls import include, path, re_path, register_converter
from django.views.generic import TemplateView

from itou.utils.urls import SiretConverter
from itou.www.dashboard import views as dashboard_views
from itou.www.login import views as login_views
from itou.www.signup import views as signup_views


register_converter(SiretConverter, "siret")

urlpatterns = [
    path("admin/", admin.site.urls),
    # --------------------------------------------------------------------------------------
    # allauth URLs. Order is important because some URLs are overriden.
    # --------------------------------------------------------------------------------------
    # Override allauth `account_signup` URL.
    # /accounts/signup/ <=> account_signup
    # We don't want any user to be able to signup using the default allauth `signup` url
    # because we have multiple specific signup processes for different kind of users.
    re_path(r"^accounts/signup/$", signup_views.signup),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_login` URL.
    # /accounts/login/ <=> account_login
    # We override this view because the login page should look slightly differently
    # for job seekers, prescribers and employers.
    # Also, PEAMU is only available for job seekers.
    re_path(r"^accounts/login/$", login_views.login),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_change_password` URL.
    # /accounts/password/change/ <=> account_change_password
    # https://github.com/pennersr/django-allauth/issues/468
    re_path(r"^accounts/password/change/$", dashboard_views.password_change),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_reset_password` URL.
    # Avoid user enumeration via password reset page.
    re_path(r"^accounts/password/reset/$", signup_views.ItouPasswordResetView.as_view()),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_logout` URL.
    # /accounts/logout/ <=> account_logout
    # We need custom code to process PEAMU logout.
    re_path(r"^accounts/logout/$", dashboard_views.logout),
    # --------------------------------------------------------------------------------------
    # Other allauth URLs.
    path("accounts/", include("allauth.urls")),
    # --------------------------------------------------------------------------------------
    # PEAMU URLs.
    path("accounts/", include("itou.allauth_adapters.peamu.urls")),
    # FranceConnect URLs.
    path("franceconnect/", include("itou.france_connect.urls")),
    # --------------------------------------------------------------------------------------
    # API.
    path("api/v1/", include("itou.api.urls", namespace="v1")),
    path("api/", include("itou.www.api.urls")),
    # www.
    path("", include("itou.www.home.urls")),
    path("apply/", include("itou.www.apply.urls")),
    path("approvals/", include("itou.www.approvals_views.urls")),
    path("autocomplete/", include("itou.www.autocomplete.urls")),
    path("dashboard/", include("itou.www.dashboard.urls")),
    path("employee_record/", include("itou.www.employee_record_views.urls")),
    path("institutions/", include("itou.www.institutions_views.urls")),
    path("invitations/", include("itou.www.invitations_views.urls")),
    path("prescribers/", include("itou.www.prescribers_views.urls")),
    path("search/", include("itou.www.search.urls")),
    path("siae/", include("itou.www.siaes_views.urls")),
    path("signup/", include("itou.www.signup.urls")),
    path("stats/", include("itou.www.stats.urls")),
    path("versions/", include("itou.www.releases.urls")),
    # Enable Anymailâ€™s status tracking
    # https://anymail.readthedocs.io/en/stable/esps/mailjet/#status-tracking-webhooks
    re_path(r"^webhooks/anymail/", include("anymail.urls")),
    path("welcoming_tour/", include("itou.www.welcoming_tour.urls")),
    # Static pages.
    path("accessibility/", TemplateView.as_view(template_name="static/accessibility.html"), name="accessibility"),
    # Errors pages.
    path("403/", TemplateView.as_view(template_name="403.html"), name="403"),
    path("404/", TemplateView.as_view(template_name="404.html"), name="404"),
    path("500/", TemplateView.as_view(template_name="500.html"), name="500"),
]

if settings.DEBUG and "debug_toolbar" in settings.INSTALLED_APPS:
    import debug_toolbar

    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
    # In the local developpemnt environment, we enable the default callback URLs
    urlpatterns = [path("", include("itou.france_connect.urls"))] + urlpatterns
