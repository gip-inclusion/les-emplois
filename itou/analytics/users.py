from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
