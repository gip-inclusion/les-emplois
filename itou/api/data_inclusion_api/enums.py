import enum


class StructureTypeStr(str, enum.Enum):
    ORGA = "orga"
    SIAE = "siae"
