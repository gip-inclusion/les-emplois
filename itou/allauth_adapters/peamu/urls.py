from django.urls import include, re_path

from itou.allauth_adapters.peamu import views as peamu_views
from itou.allauth_adapters.peamu.provider import PEAMUProvider


def default_urlpatterns(provider):
    urlpatterns = [
        re_path(r"^login/$", peamu_views.oauth2_login, name=provider.id + "_login"),
        re_path(r"^login/callback/$", peamu_views.oauth2_callback, name=provider.id + "_callback"),
    ]

    return [
        re_path(r"^profile/$", peamu_views.redirect_to_dashboard_view),
        re_path(r"^" + provider.get_slug() + r"/", include(urlpatterns)),
    ]


urlpatterns = default_urlpatterns(PEAMUProvider)
