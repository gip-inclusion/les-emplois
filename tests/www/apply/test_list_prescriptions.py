from urllib.parse import unquote

import factory
from django.urls import reverse
from pytest_django.asserts import assertContains, assertNotContains, assertNumQueries

from itou.job_applications.enums import JobApplicationState, SenderKind
from itou.job_applications.models import JobApplicationWorkflow
from itou.prescribers.enums import PrescriberOrganizationKind
from itou.utils.urls import add_url_params
from tests.job_applications.factories import JobApplicationFactory
from tests.prescribers.factories import (
    PrescriberOrganizationWith2MembershipFactory,
)
from tests.users.factories import JobSeekerFactory, PrescriberFactory
from tests.utils.htmx.test import assertSoupEqual, update_page_with_htmx
from tests.utils.test import BASE_NUM_QUERIES, assert_previous_step, parse_response_to_soup


BESOIN_DUN_CHIFFRE = "besoin-dun-chiffre"


def test_list_prescriptions(client):
    """
    Connect as Thibault to see a list of job applications
    sent by his organization (Pôle emploi).
    """
    job_application = JobApplicationFactory(sent_by_authorized_prescriber_organisation=True)
    organization = job_application.sender_prescriber_organization
    client.force_login(job_application.sender)

    response = client.get(reverse("apply:list_prescriptions"))
    assert_previous_step(response, reverse("dashboard:index"))
    # Has link to export with back_url set
    exports_link = unquote(
        add_url_params(reverse("apply:list_prescriptions_exports"), {"back_url": reverse("apply:list_prescriptions")})
    )
    assertContains(response, exports_link)

    # Has job application link with back_url set
    job_application_link = unquote(
        add_url_params(
            reverse("apply:details_for_prescriber", kwargs={"job_application_id": job_application.pk}),
            {"back_url": reverse("apply:list_prescriptions")},
        )
    )
    assertContains(response, job_application_link)

    # Has link to company with back_url set
    company_link = unquote(
        add_url_params(job_application.to_company.get_card_url(), {"back_url": reverse("apply:list_prescriptions")})
    )
    assertContains(response, company_link)

    # Count job applications used by the template
    assert len(response.context["job_applications_page"].object_list) == organization.jobapplication_set.count()


def test_list_prescriptions_as_unauthorized_prescriber(client):
    prescriber = PrescriberFactory()
    JobApplicationFactory(
        job_seeker_with_address=True,
        job_seeker__first_name="Supersecretname",
        job_seeker__last_name="Unknown",
        job_seeker__created_by=PrescriberFactory(),  # to check for useless queries
        sender=prescriber,
        sender_kind=SenderKind.PRESCRIBER,
    )
    client.force_login(prescriber)

    with assertNumQueries(
        BASE_NUM_QUERIES
        + 1  # fetch django session
        + 1  # fetch user
        + 1  # fetch user memberships
        + 1  # get list of senders (distinct sender_id)
        + 1  # get list of job seekers (distinct job_seeker_id)
        + 1  # get list of administrative criteria
        + 2  # get list of job application + prefetch of job descriptions
        + 1  # get list of siaes (distinct to_company_id)
        + 3  # count, list & prefetch of job application
        + 1  # get job seekers approvals
        + 1  # check user authorized membership (can_edit_personal_information)
        + 3  # get job seekers administrative criteria
        + 3  # update session
    ):
        response = client.get(reverse("apply:list_prescriptions"))

    assertContains(response, "<h3>S… U…</h3>", html=True)
    # Unfortunately, the job seeker's name is available in the filters
    # assertNotContains(response, "Supersecretname")


def test_list_prescriptions_filtered_by_state(client):
    """
    Thibault wants to filter a list of job applications
    by the default initial state.
    """
    prescriber = PrescriberFactory()
    job_application, *others = JobApplicationFactory.create_batch(
        3, sender=prescriber, state=factory.Iterator(JobApplicationWorkflow.states)
    )
    client.force_login(prescriber)

    response = client.get(reverse("apply:list_prescriptions"), {"states": [job_application.state]})
    applications = response.context["job_applications_page"].object_list
    assert len(applications) == 1
    assert applications[0].state == job_application.state

    response = client.get(reverse("apply:list_prescriptions"))
    applications = response.context["job_applications_page"].object_list
    assert len(applications) == 3


def test_list_prescriptions_filtered_by_sender(client):
    organization = PrescriberOrganizationWith2MembershipFactory()
    a_prescriber, another_prescriber = organization.members.all()
    JobApplicationFactory(sender=another_prescriber, sender_prescriber_organization=organization)
    client.force_login(a_prescriber)

    response = client.get(reverse("apply:list_prescriptions"), {"senders": another_prescriber.pk})
    applications = response.context["job_applications_page"].object_list
    assert len(applications) == 1
    assert applications[0].sender.id == another_prescriber.pk


def test_list_prescriptions_filtered_by_job_seeker(client):
    job_seeker = JobSeekerFactory()
    prescriber = PrescriberFactory()
    JobApplicationFactory(sender=prescriber, job_seeker=job_seeker)
    JobApplicationFactory.create_batch(2, sender=prescriber)
    client.force_login(prescriber)

    response = client.get(reverse("apply:list_prescriptions"), {"job_seekers": [job_seeker.pk]})
    applications = response.context["job_applications_page"].object_list
    assert len(applications) == 1
    assert applications[0].job_seeker.pk == job_seeker.pk

    response = client.get(reverse("apply:list_prescriptions"))
    applications = response.context["job_applications_page"].object_list
    assert len(applications) == 3


def test_list_prescriptions_filtered_by_company(client):
    prescriber = PrescriberFactory()
    job_application, *others = JobApplicationFactory.create_batch(3, sender=prescriber)
    client.force_login(prescriber)

    response = client.get(reverse("apply:list_prescriptions"), {"to_companies": [job_application.to_company.pk]})
    applications = response.context["job_applications_page"].object_list
    assert len(applications) == 1
    assert applications[0].to_company.pk == job_application.to_company.pk

    response = client.get(reverse("apply:list_prescriptions"))
    applications = response.context["job_applications_page"].object_list
    assert len(applications) == 3


def test_list_prescriptions_filters(client, snapshot):
    client.force_login(PrescriberFactory())

    response = client.get(reverse("apply:list_prescriptions"))
    assert response.status_code == 200
    filter_form = parse_response_to_soup(response, "#offcanvasApplyFilters")
    assert str(filter_form) == snapshot()


def test_list_prescriptions_htmx_filters(client):
    prescriber = PrescriberFactory()
    JobApplicationFactory(sender=prescriber, state=JobApplicationState.ACCEPTED)
    client.force_login(prescriber)

    url = reverse("apply:list_prescriptions")
    response = client.get(url)
    page = parse_response_to_soup(response, selector="#main")
    # Simulate the data-sync-with and check both checkboxes.
    refused_checkboxes = page.find_all(
        "input",
        attrs={"name": "states", "value": "refused"},
    )
    assert len(refused_checkboxes) == 2
    for refused_checkbox in refused_checkboxes:
        refused_checkbox["checked"] = ""

    response = client.get(
        url,
        {"states": ["refused"]},
        headers={"HX-Request": "true"},
    )
    update_page_with_htmx(page, f"form[hx-get='{url}']", response)

    response = client.get(url, {"states": ["refused"]})
    fresh_page = parse_response_to_soup(response, selector="#main")
    assertSoupEqual(page, fresh_page)


def test_list_prescriptions_exports_without_organization(client):
    client.force_login(PrescriberFactory())

    response = client.get(reverse("apply:list_prescriptions_exports"))
    assert_previous_step(response, reverse("dashboard:index"))
    assertNotContains(response, BESOIN_DUN_CHIFFRE)


def test_list_prescriptions_exports_with_organization(client):
    client.force_login(PrescriberFactory(membership=True))

    response = client.get(reverse("apply:list_prescriptions_exports"))
    assert_previous_step(response, reverse("dashboard:index"))
    assertNotContains(response, BESOIN_DUN_CHIFFRE)


def test_list_prescriptions_exports_as_pole_emploi_prescriber(client, snapshot):
    job_application = JobApplicationFactory(
        sent_by_authorized_prescriber_organisation=True,
        sender_prescriber_organization__kind=PrescriberOrganizationKind.PE,
    )
    client.force_login(job_application.sender)

    response = client.get(reverse("apply:list_prescriptions_exports"))
    assert_previous_step(response, reverse("dashboard:index"))
    assertContains(response, "Toutes les candidatures")
    soup = parse_response_to_soup(response, selector=f"#{BESOIN_DUN_CHIFFRE}")
    assert str(soup) == snapshot


def test_list_prescriptions_exports_as_employer(client):
    job_application = JobApplicationFactory()
    client.force_login(job_application.to_company.members.get())

    response = client.get(reverse("apply:list_prescriptions_exports"))
    assertNotContains(response, BESOIN_DUN_CHIFFRE)


def test_list_prescriptions_exports_back_to_list(client):
    client.force_login(PrescriberFactory())

    response = client.get(
        add_url_params(reverse("apply:list_prescriptions_exports"), {"back_url": reverse("apply:list_prescriptions")})
    )
    assert_previous_step(response, reverse("apply:list_prescriptions"), back_to_list=True)
    assertNotContains(response, BESOIN_DUN_CHIFFRE)


def test_list_prescriptions_exports_download(client):
    job_application = JobApplicationFactory()
    client.force_login(job_application.sender)

    response = client.get(reverse("apply:list_prescriptions_exports_download"))
    assert 200 == response.status_code
    assert "spreadsheetml" in response.get("Content-Type")


def test_list_prescriptions_exports_download_as_employer(client):
    job_application = JobApplicationFactory()
    client.force_login(job_application.to_company.members.get())

    response = client.get(reverse("apply:list_prescriptions_exports_download"))
    assert 200 == response.status_code
    assert "spreadsheetml" in response.get("Content-Type")


def test_list_prescriptions_exports_download_by_month(client):
    job_application = JobApplicationFactory()
    client.force_login(job_application.sender)

    response = client.get(
        reverse(
            "apply:list_prescriptions_exports_download",
            kwargs={"month_identifier": job_application.created_at.strftime("%Y-%d")},
        )
    )
    assert 200 == response.status_code
    assert "spreadsheetml" in response.get("Content-Type")
