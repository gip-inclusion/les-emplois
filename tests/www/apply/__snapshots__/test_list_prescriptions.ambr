# serializer version: 1
# name: test_list_prescriptions_exports_as_pole_emploi_prescriber
  '''
  <div class="c-info mb-3 mb-md-4" id="besoin-dun-chiffre">
                              <button aria-controls="collapseBesoinChiffre" aria-expanded="false" class="c-info__summary collapsed" data-bs-target="#collapseBesoinChiffre" data-bs-toggle="collapse">
                                  <span>Besoin d'un chiffre ?</span>
                              </button>
                              
                                  <div class="c-info__detail collapse" id="collapseBesoinChiffre">
                                      Accédez aux données de votre agence (non nominatives) compilées, calculées et mises à jour quotidiennement :
                                      <ul class="mb-0">
                                          <li>
                                              <a href="/stats/pe/state/main" rel="noopener" target="_blank">Voir les données de l'ensemble de l'état des candidatures orientées</a>
                                          </li>
                                          <li>
                                              <a href="/stats/pe/conversion/main" rel="noopener" target="_blank">Voir les données du taux de transformation des candidatures</a>
                                          </li>
                                      </ul>
                                  </div>
                              
                          </div>
  '''
# ---
# name: test_list_prescriptions_filters
  '''
  <div class="c-aside-filters__card collapse show" id="asideFiltersCollapse">
                              <form hx-get="/apply/prescriptions/list" hx-indicator="#job-applications-section" hx-push-url="true" hx-swap="outerHTML" hx-target="#job-applications-section" hx-trigger="change delay:.5s, duetChange delay:.5s">
                                  <div class="c-aside-filters__card__body">
                                      <fieldset>
      <legend>Statut candidature</legend>
      <ul>
          
              <li>
                  <div class="form-check">
                      <input class="form-check-input" id="id_states_0" name="states" type="checkbox" value="new"/>
                      <label class="form-check-label" for="id_states_0">Nouvelle candidature</label>
                  </div>
              </li>
          
              <li>
                  <div class="form-check">
                      <input class="form-check-input" id="id_states_1" name="states" type="checkbox" value="processing"/>
                      <label class="form-check-label" for="id_states_1">Candidature à l'étude</label>
                  </div>
              </li>
          
              <li>
                  <div class="form-check">
                      <input class="form-check-input" id="id_states_2" name="states" type="checkbox" value="postponed"/>
                      <label class="form-check-label" for="id_states_2">Candidature en attente</label>
                  </div>
              </li>
          
              <li>
                  <div class="form-check">
                      <input class="form-check-input" id="id_states_3" name="states" type="checkbox" value="prior_to_hire"/>
                      <label class="form-check-label" for="id_states_3">Action préalable à l’embauche</label>
                  </div>
              </li>
          
              <li>
                  <div class="form-check">
                      <input class="form-check-input" id="id_states_4" name="states" type="checkbox" value="accepted"/>
                      <label class="form-check-label" for="id_states_4">Candidature acceptée</label>
                  </div>
              </li>
          
              <li>
                  <div class="form-check">
                      <input class="form-check-input" id="id_states_5" name="states" type="checkbox" value="refused"/>
                      <label class="form-check-label" for="id_states_5">Candidature déclinée</label>
                  </div>
              </li>
          
              <li>
                  <div class="form-check">
                      <input class="form-check-input" id="id_states_6" name="states" type="checkbox" value="cancelled"/>
                      <label class="form-check-label" for="id_states_6">Embauche annulée</label>
                  </div>
              </li>
          
              <li>
                  <div class="form-check">
                      <input class="form-check-input" id="id_states_7" name="states" type="checkbox" value="obsolete"/>
                      <label class="form-check-label" for="id_states_7">Embauché ailleurs</label>
                  </div>
              </li>
          
      </ul>
  </fieldset>
  
                                      
  
  
      <hr/>
      <fieldset>
          <div class="form-group"><label class="form-label" for="id_job_seekers">Nom du candidat</label><select class="form-select django-select2" data-allow-clear="true" data-minimum-input-length="0" data-placeholder="" data-theme="bootstrap-5" id="id_job_seekers" lang="fr" multiple="" name="job_seekers">
  </select></div>
      </fieldset>
  
  
                                      
  
  
      <hr/>
      <fieldset>
          <legend>Structure destinataire</legend>
          <div class="form-group"><label class="visually-hidden" for="id_to_companies">Structure</label><select class="form-select django-select2" data-allow-clear="true" data-minimum-input-length="0" data-placeholder="" data-theme="bootstrap-5" id="id_to_companies" lang="fr" multiple="" name="to_companies">
  </select></div>
      </fieldset>
  
  
  
                                      
  
  
      <hr/>
      <fieldset>
          <legend>Émetteur</legend>
  
          
              <div class="form-group"><label class="form-label" for="id_senders">Nom de la personne</label><select class="form-select django-select2" data-allow-clear="true" data-minimum-input-length="0" data-placeholder="" data-theme="bootstrap-5" id="id_senders" lang="fr" multiple="" name="senders">
  </select></div>
          
  
          
  
          
      </fieldset>
  
  
                                      
  
  <hr/>
  <fieldset>
      <legend>Date d'envoi de la candidature</legend>
      <div class="form-group"><label class="form-label" for="id_start_date">À partir du</label><duet-date-picker class="" identifier="id_start_date" name="start_date" placeholder="À partir du"></duet-date-picker></div>
      <div class="form-group"><label class="form-label" for="id_end_date">Jusqu'au</label><duet-date-picker class="" identifier="id_end_date" name="end_date" placeholder="Jusqu'au"></duet-date-picker></div>
  </fieldset>
  
                                      
  
  
      <hr/>
      <fieldset>
          <div class="form-group"><label class="form-label" for="id_selected_jobs">Fiches de poste</label><select class="form-select django-select2" data-allow-clear="true" data-minimum-input-length="0" data-placeholder="" data-theme="bootstrap-5" id="id_selected_jobs" lang="fr" multiple="" name="selected_jobs">
  </select></div>
      </fieldset>
  
  
                                      
  
  
      <hr/>
      <fieldset>
          <legend>Statut du PASS IAE</legend>
          <div class="mb-2"><div class="form-check"><input class="form-check-input" id="id_pass_iae_active" name="pass_iae_active" type="checkbox"/><label class="form-check-label" for="id_pass_iae_active">Actif</label></div></div>
          <div class=""><div class="form-check"><input class="form-check-input" id="id_pass_iae_suspended" name="pass_iae_suspended" type="checkbox"/><label class="form-check-label" for="id_pass_iae_suspended">Suspendu</label></div></div>
      </fieldset>
  
  
                                      
  
  
      <hr/>
      <fieldset>
          <legend>Éligibilité IAE</legend>
          <div class="form-group"><div class="form-check"><input class="form-check-input" id="id_eligibility_validated" name="eligibility_validated" type="checkbox"/><label class="form-check-label" for="id_eligibility_validated">Éligibilité validée</label></div></div>
      </fieldset>
  
  
                                      
  
  
  
      <hr/>
      <fieldset>
          <legend>Critères administratifs déclarés</legend>
          <div class=""><select class="form-select django-select2" data-allow-clear="true" data-minimum-input-length="0" data-placeholder="" data-theme="bootstrap-5" id="id_criteria" lang="fr" multiple="" name="criteria">
    <option value="1">Bénéficiaire du RSA</option>
  
    <option value="2">Allocataire ASS</option>
  
    <option value="3">Allocataire AAH</option>
  
    <option value="4">DETLD (+ 24 mois)</option>
  
    <option value="5">Niveau d'étude 3 (CAP, BEP) ou infra</option>
  
    <option value="6">Senior (+50 ans)</option>
  
    <option value="7">Jeune (-26 ans)</option>
  
    <option value="8">Sortant de l'ASE</option>
  
    <option value="9">DELD (12-24 mois)</option>
  
    <option value="10">Travailleur handicapé</option>
  
    <option value="11">Parent isolé</option>
  
    <option value="12">Personne sans hébergement ou hébergée ou ayant un parcours de rue</option>
  
    <option value="13">Réfugié statutaire, bénéficiaire d'une protection temporaire, protégé subsidiaire ou demandeur d'asile</option>
  
    <option value="14">Résident ZRR</option>
  
    <option value="15">Résident QPV</option>
  
    <option value="16">Sortant de détention ou personne placée sous main de justice</option>
  
    <option value="17">Maîtrise de la langue française</option>
  
    <option value="18">Mobilité</option>
  
  </select></div>
      </fieldset>
  
  
                                      
  
  
      <hr/>
      <fieldset>
          <div class="form-group"><label class="form-label" for="id_departments">Département du candidat</label><select class="form-select django-select2" data-allow-clear="true" data-minimum-input-length="0" data-placeholder="" data-theme="bootstrap-5" id="id_departments" lang="fr" multiple="" name="departments">
  </select></div>
      </fieldset>
  
  
                                  </div>
                              </form>
                          </div>
  '''
# ---
